src/libcore/cmp.rs

```diff
@@ -32,6 +31,66 @@
 //! assert_eq!(x == y, false);
 //! assert_eq!(x.eq(&y), false);
 //! ```
+//!
+//! This module also provides the following two traits:
+//!
+//! * [`Eq`]: which requires `PartialEq`, states that all values of a type can be
+//!   compared to all others,
+//! * [`Ord`]: which requires `PartialOrd` and `Eq`, states that the ordering
+//!   relations expressed by the comparison operators form all total ordering relations 
+//!   (more on this below).
+//!
+//! The distinction between these traits is important: it tell us what it _means_ for a 
+//! sequence of values to be ordered.
+//!
+//! Two commonly encountered examples are integers and floating-point numbers. For example, every [`i32`] 
+//! value is comparable to every other. If you can tell that two `i32` values are distinc, then these two
+//! values never compare equal, and if you sorte a sequence of `i32` values like `[1, 1, 0, -5]` using `<=`,
+//! doing so will always produce the same unique result: `[-5, 0, 1, 1]`. We say that `<=` imposes a _total_
+//! order, and denote this by implementing `Ord` for `i32`.
+//!
+//! However, none of this holds for IEEE-754 floating-point numbers like `f32`. There is a special `f32` value 
+//! called "Not a Number" (`NaN`) for which any relational operation alwys return false, that is: 
+//! `a == NaN` is always `false`, even when `a` is `NaN`. For this reason `f32` only implements `PartialOrd`, and it 
+//! does not implement neither `Eq` nor `Ord`. What does this mean in practice? It means two things:
+//!
+//! * there are sequences of `f32`s that we can order using `<=` and will always produce the same unique 
+//!   result: sorting `[3.0, 5.0, 1.0]` into `[1.0, 3.0, 5.0]`.
+//!
+//! * there are also sequences of `f32`s that when ordered using `<=` will produce a different result depending
+//!  on which algorithm we use to sort them: we can sort `[-0.0, NaN, +0.0] into all its permutations 
+//! (`[-0.0, +0.0, NaN]`, `[+0.0, -0.0, NaN]`, `[NaN, +0.0, -0.0]`, etc. are all "sorted" under `<=`).
```

Comment:
I don't really like this part (the two bullet points) :/

I think I would replace everything from "What does this mean in practice?" to "are all "sorted" under `<=`)." with something like:

> While this is no problem for sequences like `[3.0, 5.0, 1.0]`, there is no clear order for sequences containing `NaN`: what is the correct way to sort `[-1.0, NaN, 4.0]`? The order `[NaN, -1.0, 4.0]` would be exactly as correct as `[-1.0, 4.0, NaN]`. So there is no unique ordering for floats.

Also: "are all "sorted" under `<=`" -> is that even correct? I'm not quite sure what you mean, but `NaN <= +0.0` is `false`.

Reply:
Yes I wasn't fully happy with this either. For me, all are sorted under `<=`  means that `!(a > b)` is true for all adjacent pairs in the sequence. When either `a` or `b` are `NaN` that's automatically satisfied. Sadly for floats that's not exactly the same as `a <= b` being `true` for all adjacent pairs...